src/wasm_udf/wasm.rs

```diff
@@ -11,22 +15,518 @@ use datafusion::{
 use datafusion::prelude::*;
 use datafusion::{error::Result, physical_plan::functions::make_scalar_function};
 
+use wasmtime::{Engine, Instance, Memory, Module, Store, TypedFunc, Val, ValType};
+
+use super::data_types::{get_wasm_type, CreateFunctionDataType, CreateFunctionLanguage};
+
+use wasi_common::WasiCtx;
+use wasmtime_wasi::sync::WasiCtxBuilder;
+
 use std::sync::Arc;
-use wasmtime::{Instance, Module, Store, Val, ValType};
+use std::vec;
+
+extern crate rmp_serde;
+extern crate serde;
+use rmp_serde::Serializer;
+
+use serde::Serialize;
 
-fn wasm_type_to_arrow_type(t: &ValType) -> Result<DataType> {
+use rmpv::Value;
+
+fn sql_type_to_arrow_type(t: &CreateFunctionDataType) -> Result<DataType> {
     match t {
-        ValType::I32 => Ok(DataType::Int32),
-        ValType::I64 => Ok(DataType::Int64),
-        ValType::F32 => Ok(DataType::Float32),
-        ValType::F64 => Ok(DataType::Float64),
-        _ => Err(DataFusionError::Internal(format!(
-            "Unsupported WASM type: {:?}",
-            t
+        // legacy WASM-native type names
+        CreateFunctionDataType::I32 => Ok(DataType::Int32),
+        CreateFunctionDataType::I64 => Ok(DataType::Int64),
+        CreateFunctionDataType::F32 => Ok(DataType::Float32),
+        CreateFunctionDataType::F64 => Ok(DataType::Float64),
+        // DDL type names
+        CreateFunctionDataType::SMALLINT => Ok(DataType::Int16),
+        CreateFunctionDataType::INT => Ok(DataType::Int32),
+        CreateFunctionDataType::BIGINT => Ok(DataType::Int64),
+        CreateFunctionDataType::CHAR => Ok(DataType::Utf8),
+        CreateFunctionDataType::VARCHAR => Ok(DataType::Utf8),
+        CreateFunctionDataType::TEXT => Ok(DataType::Utf8),
+        CreateFunctionDataType::DECIMAL { precision, scale } => {
+            Ok(DataType::Decimal128(*precision, *scale))
+        }
+        CreateFunctionDataType::FLOAT => Ok(DataType::Float32),
+        CreateFunctionDataType::REAL => Ok(DataType::Float32),
+        CreateFunctionDataType::DOUBLE => Ok(DataType::Float64),
+        CreateFunctionDataType::BOOLEAN => Ok(DataType::Boolean),
+        CreateFunctionDataType::DATE => Ok(DataType::Date32),
+        CreateFunctionDataType::TIMESTAMP => {
+            Ok(DataType::Timestamp(TimeUnit::Nanosecond, None))
+        }
+    }
+}
+
+fn get_wasm_module_exported_fn<Params, Results>(
+    instance: &Instance,
+    store: &mut Store<WasiCtx>,
+    export_name: &str,
+) -> anyhow::Result<TypedFunc<Params, Results>>
```

Comment:
Is the `anyhow` crate necessary here at all? In other places, we either return a `DataFusionError` directly or define a specific error type, e.g. https://github.com/splitgraph/seafowl/blob/main/src/frontend/http_utils.rs#L49-L116

Reply:
Good point, got rid of `anyhow`